include/dice/sparse_map/internal/sparse_hash.hpp

```diff
@@ -0,0 +1,943 @@
+				assert(load_threshold_rehash_ >= size());
+
+				if (nb_elements_insert > 0 &&
+					nb_free_buckets < size_type(nb_elements_insert)) {
+					reserve(size() + size_type(nb_elements_insert));
+				}
+			}
+
+			for (; first != last; ++first) {
+				insert(*first);
+			}
+		}
+
+		template<typename K, typename M>
+		std::pair<iterator, bool> insert_or_assign(K &&key, M &&obj) {
+			auto it = try_emplace(std::forward<K>(key), std::forward<M>(obj));
+			if (!it.second) {
+				it.first->second = std::forward<M>(obj);
+			}
+
+			return it;
+		}
+
+		template<typename K, typename M>
+		iterator insert_or_assign(const_iterator hint, K &&key, M &&obj) {
+			if (hint != cend() && keq_(KeyValueSelect::key(*hint), key)) {
+				auto it = mutable_iterator(hint);
+				it->second = std::forward<M>(obj);
+
+				return it;
+			}
+
+			return insert_or_assign(std::forward<K>(key), std::forward<M>(obj)).first;
+		}
+
+		template<typename ...Args>
+		std::pair<iterator, bool> emplace(Args &&...args) {
+			return insert(value_type(std::forward<Args>(args)...));
+		}
+
+		template<typename ...Args>
+		iterator emplace_hint(const_iterator hint, Args &&...args) {
+			return insert_hint(hint, value_type(std::forward<Args>(args)...));
+		}
+
+		template<typename K, typename ...Args>
+		std::pair<iterator, bool> try_emplace(K &&key, Args &&...args) {
+			return insert_impl(key, std::piecewise_construct,
+							   std::forward_as_tuple(std::forward<K>(key)),
+							   std::forward_as_tuple(std::forward<Args>(args)...));
+		}
+
+		template<typename K, typename ...Args>
+		iterator try_emplace_hint(const_iterator hint, K &&key, Args &&...args) {
+			if (hint != cend() && keq_(KeyValueSelect::key(*hint), key)) {
+				return mutable_iterator(hint);
+			}
+
+			return try_emplace(std::forward<K>(key), std::forward<Args>(args)...).first;
+		}
+
+		/**
+		 * Here to avoid `template<class K> size_type erase(const K& key)` being used
+		 * when we use an iterator instead of a const_iterator.
+		 */
+		iterator erase(iterator pos) {
+			assert(pos != end() && n_elements_ > 0);
+			//vector iterator with fancy pointers have a problem with ->
+			auto next_bucket_it = pos.cur_bucket_->erase(buckets_.element_allocator(), pos.bucket_it_);
+			n_elements_--;
+			n_deleted_elements_++;
+
+			if (next_bucket_it != pos.cur_bucket_->end()) {
+				return iterator{pos.cur_bucket_,
+								buckets_.end(),
+								next_bucket_it};
+			}
+
+			auto it_sparse_buckets_next = pos.cur_bucket_;
+			do {
+				++it_sparse_buckets_next;
+			} while (it_sparse_buckets_next != buckets_.end()
+					 && (*it_sparse_buckets_next).empty());
+
+			if (it_sparse_buckets_next == buckets_.end()) {
+				return end();
+			} else {
+				return iterator{it_sparse_buckets_next,
+								buckets_.end(),
+								it_sparse_buckets_next->begin()};
+			}
+		}
+
+		iterator erase(const_iterator pos) {
+			return erase(mutable_iterator(pos));
+		}
+
+		iterator erase(const_iterator first, const_iterator last) {
```

Comment:
This whole function doesn't make much sense to me. Erase can trigger a rehash, right? So we cannot even be sure that the range of entries that the user could observe before the call between first and last is even correctly erase.

Reply:
Yeah I was wondering the same thing. I honestly don't know if erasing ranges from hash maps ever makes sense, I mean you don't even know whats in the range in the first place as its effectively random.

Just remove?

Reply:
I would suggest to keep the function but deactivate it with an requires false + an explanation why it is not supported ATM.

Reply:
Actually no, it cannot trigger a rehash, because you couldn't do erase loops if it did.
But the function is still pretty useless I would say.